File: 0278-first-bad-version/0278-first-bad-version.c
// The API isBadVersion is defined for you.
// bool isBadVersion(int version);

int firstBadVersion(int n) {
    long    l = 0,
            r = n;
    while (l < r) {
        int m = l + (r - l)/2;
        
        if (isBadVersion(m)) {
            r = m;
        } else {
            l = m + 1;
        }
    }
    return l;
}

File: 0278-first-bad-version/0278-first-bad-version.rs
// The API isBadVersion is defined for you.
// isBadVersion(version:i32)-> bool;
// to call it use self.isBadVersion(version)

impl Solution {
    pub fn first_bad_version(&self, n: i32) -> i32 {
		let (mut l, mut r) = (0, n);
        
        while l < r {
            if self.isBadVersion(l + (r - l)/2) {
                r = l + (r - l)/2
            } else {
                l = l + (r - l)/2 + 1
            }
        }
        
        l
    }
}